### PR TITLE
[INLONG-1796]DataProxy support monitor indicator with JMX. #1796

### LINKED ISSUE #1796: [Feature]DataProxy support monitor indicator with JMX.
### Description

[Feature]DataProxy provide monitor indicator based on JMX, user can implement the code that read the metrics and report to user-defined monitor system.

DataProxy provide monitor indicator based on JMX, user can implement the code that read the metrics and report to user-defined monitor system.
Source-module and Sink-module can add monitor metric class that is the subclass of org.apache.inlong.commons.config.metrics.MetricItemSet, and register it to MBeanServer.
User-defined plugin can get module metric with JMX, and report metric data to different monitor system.

User can describe the configuration in the file "common.properties ".
For example:
`
metricDomains=DataProxy
metricDomains.DataProxy.domainListeners=com.tencent.pcg.atta.dataproxy.metrics.m007.M007MetricListener org.apache.inlong.dataproxy.metrics.prometheus.PrometheusMetricListener
metricDomains.DataProxy.snapshotInterval=60000
`

The JMX domain name of DataProxy is "DataProxy". 
It is defined by the parameter "metricDomains".
The listeners of JMX domain is defined by the parameter "metricDomains.$domainName.domainListeners".
The class names of the listeners is separated by the space char. 
The listener class need to implement the interface "org.apache.inlong.dataproxy.metrics.MetricListener".
The snapshot interval of the listeners is defined by the parameter "metricDomains.$domainName.snapshotInterval", the parameter unit is "millisecond".

The method proto of org.apache.inlong.dataproxy.metrics.MetricListener is:
public void snapshot(String domain, List<MetricItemValue> itemValues);

The field of MetricItemValue.dimensions has these key(The fields of DataProxyMetricItem defined by the Annotation "@Dimension"):

- public String clusterId;
- public String sourceId;
- public String sourceDataId;
- public String inlongGroupId;
- public String inlongStreamId;
- public String sinkId;
- public String sinkDataId;

The field of MetricItemValue.metrics has these key(The fields of DataProxyMetricItem defined by the Annotation "@CountMetric"):

- readSuccessCount
- readSuccessSize
- readSuccessCount
- readSuccessSize
- readFailCount
- readFailSize
- sendCount
- sendSize
- sendSuccessCount
- sendSuccessSize
- sendFailCount
- sendFailSize
- sinkDuration, the unit is millisecond, the duration is between current timepoint and the timepoint in sending to sink destination.
- nodeDuration, the unit is millisecond, the duration is between current timepoint and the timepoint in getting event from source.
- wholeDuration, the unit is millisecond, the duration is between current timepoint and the timepoint in generating event.

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by luchunliang**
Fixes #1796

### Motivation

[Feature]DataProxy provide monitor indicator based on JMX, user can implement the code that read the metrics and report to user-defined monitor system.

DataProxy provide monitor indicator based on JMX, user can implement the code that read the metrics and report to user-defined monitor system.
Source-module and Sink-module can add monitor metric class that is the subclass of org.apache.inlong.commons.config.metrics.MetricItemSet, and register it to MBeanServer.
User-defined plugin can get module metric with JMX, and report metric data to different monitor system.

User can describe the configuration in the file "common.properties ".
For example:
`
metricDomains=DataProxy
metricDomains.DataProxy.domainListeners=com.tencent.pcg.atta.dataproxy.metrics.m007.M007MetricListener org.apache.inlong.dataproxy.metrics.prometheus.PrometheusMetricListener
metricDomains.DataProxy.snapshotInterval=60000
`

The JMX domain name of DataProxy is "DataProxy". 
It is defined by the parameter "metricDomains".
The listeners of JMX domain is defined by the parameter "metricDomains.$domainName.domainListeners".
The class names of the listeners is separated by the space char. 
The listener class need to implement the interface "org.apache.inlong.dataproxy.metrics.MetricListener".
The snapshot interval of the listeners is defined by the parameter "metricDomains.$domainName.snapshotInterval", the parameter unit is "millisecond".

The method proto of org.apache.inlong.dataproxy.metrics.MetricListener is:
public void snapshot(String domain, List<MetricItemValue> itemValues);

The field of MetricItemValue.dimensions has these key(The fields of DataProxyMetricItem defined by the Annotation "@Dimension"):

- public String clusterId;
- public String sourceId;
- public String sourceDataId;
- public String inlongGroupId;
- public String inlongStreamId;
- public String sinkId;
- public String sinkDataId;

The field of MetricItemValue.metrics has these key(The fields of DataProxyMetricItem defined by the Annotation "@CountMetric"):

- readSuccessCount
- readSuccessSize
- readSuccessCount
- readSuccessSize
- readFailCount
- readFailSize
- sendCount
- sendSize
- sendSuccessCount
- sendSuccessSize
- sendFailCount
- sendFailSize
- sinkDuration, the unit is millisecond, the duration is between current timepoint and the timepoint in sending to sink destination.
- nodeDuration, the unit is millisecond, the duration is between current timepoint and the timepoint in getting event from source.
- wholeDuration, the unit is millisecond, the duration is between current timepoint and the timepoint in generating event.

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by dockerzhang** [comment]
@luchunliang please add some UTs to verify this feature.

**Review by gosonzhang — APPROVED** [review]
+1

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * start
+     */
+    public void start() {
+        try {
+            this.reload();
+            this.setReloadTimer();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (inter12) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java:84]: should not print stacktrace direct

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            LOG.error(e.getMessage(), e);
+        }
+    }
+
+    /**
+     * setReloadTimer
+     */
+    private void setReloadTimer() {
+        reloadTimer = new Timer(true);
+        TimerTask task = new TimerTask() {
```
Comment (inter12) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java:105]: TimerTask is not recommend to use . use ScheduledExecutorService instead

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/IdTopicConfigHolder.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * start
+     */
+    public void start() {
+        try {
+            this.reload();
+            this.setReloadTimer();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (inter12) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/IdTopicConfigHolder.java:87]: e.printstacktrace is not suitable

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ClassResourceCommonPropertiesLoader.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * loadProperties
+     * 
+     * @param  fileName
+     * @return
+     */
+    protected Map<String, String> loadProperties(String fileName) {
+        Map<String, String> result = new ConcurrentHashMap<>();
+        InputStream inStream = null;
```
Comment (inter12) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ClassResourceCommonPropertiesLoader.java:57]: recommend to use try-resource

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextCacheClusterConfigLoader.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    private Context context;
+
+    /**
+     * load
+     * 
+     * @return
+     */
+    @Override
+    public List<CacheClusterConfig> load() {
+        String clusterNames = context.getString("cacheClusterConfig");
```
Comment (inter12) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextCacheClusterConfigLoader.java:46]: extract  to constant

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextIdTopicConfigLoader.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * load
+     * 
+     * @return
+     */
+    @Override
+    public List<IdTopicConfig> load() {
+        Map<String, String> idTopicMap = context.getSubProperties("idTopicConfig.");
+        List<IdTopicConfig> configList = new ArrayList<>(idTopicMap.size());
```
Comment (inter12) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextIdTopicConfigLoader.java:44]: NPE check？

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricObserver.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * 
+ * MetricObserver
+ */
+public class MetricObserver {
+
+    public static final Logger LOG = LoggerFactory.getLogger(MetricObserver.class);
+    private static final AtomicBoolean isInited = new AtomicBoolean(false);
+    private static ScheduledExecutorService statExecutor = Executors.newScheduledThreadPool(5);
```
Comment (inter12) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricObserver.java:42]: should init with thead name for online trouble shooting

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricListenerRunnable.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    /**
+     * run
+     */
+    @Override
+    public void run() {
+        LOG.info("begin to snapshot metric:{}", domain);
+        try {
+            List<MetricItemValue> itemValues = this.getItemValues();
+            LOG.info("snapshot metric:{},size:{}", domain, itemValues.size());
+//            LOG.info("snapshot metric:{},size:{},contents:{}",
```
Comment (aloyszhang) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricListenerRunnable.java:74]: remove un-used code

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/sink/pulsar/federation/PulsarFederationSink.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+                DataProxyMetricItem metricItem = this.context.getMetricItemSet().findMetricItem(dimensions);
+                metricItem.readFailCount.incrementAndGet();
+            } catch (Throwable e) {
+                LOG.error("Channel take transaction rollback exception:" + getName(), e);
+            }
+            return Status.BACKOFF;
+        } finally {
+            tx.close();
+        }
+//        Channel channel = getChannel();
```
Comment (aloyszhang) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/sink/pulsar/federation/PulsarFederationSink.java:139]: remove un-used code

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java (hunk 2 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java:105]: ScheduledExecutorService is used to ThreadPool case, it is better for a lot of same task.
The reload task of CacheClusterConfigHolder is a single schedule task, single thread of Timer is better.

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/IdTopicConfigHolder.java (hunk 3 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/IdTopicConfigHolder.java:87]: fixed, remove this line.

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ClassResourceCommonPropertiesLoader.java (hunk 4 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ClassResourceCommonPropertiesLoader.java:57]: fixed, use try-with-resource mode.

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextIdTopicConfigLoader.java (hunk 6 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextIdTopicConfigLoader.java:44]: context.getSubProperties() should not return null.
  public ImmutableMap<String, String> getSubProperties(String prefix) {
    Preconditions.checkArgument(prefix.endsWith("."),
        "The given prefix does not end with a period (" + prefix + ")");
    Map<String, String> result = Maps.newHashMap();
    synchronized (parameters) {
      for (Entry<String, String> entry : parameters.entrySet()) {
        String key = entry.getKey();
        if (key.startsWith(prefix)) {
          String name = key.substring(prefix.length());
          result.put(name, entry.getValue());
        }
      }
    }
    return ImmutableMap.copyOf(result);
  }

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextCacheClusterConfigLoader.java (hunk 5 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/loader/ContextCacheClusterConfigLoader.java]: fixed, extract to constant.

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricObserver.java (hunk 7 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricObserver.java:42]: statExecutor is a thread pool. Runtime exception will print stack trace to log file.
It is enough to online trouble shooting based on log file.

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricListenerRunnable.java (hunk 8 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/MetricListenerRunnable.java:74]: fixed, remove un-used code.

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/sink/pulsar/federation/PulsarFederationSink.java (hunk 9 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/sink/pulsar/federation/PulsarFederationSink.java:139]: fixed, remove un-used code.

**Review by dockerzhang — APPROVED** [review]
LGTM

**Review by gosonzhang — APPROVED** [review]
+1

**Review by aloyszhang — APPROVED** [review]
+1

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java (hunk 1 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/config/holder/CacheClusterConfigHolder.java]: fixed, print exception with log4j.

**Pull request merged**